proposals/0000-autolinking-local-modules.md

```diff
@@ -0,0 +1,83 @@
+4. The workflow for creating new architecture modules without creating packages is not streamlined,
+5. It's challenging to open-source the local module, as it requires creating a dedicated library setup that adds extra friction to the process.
+
+## Detailed design
+
+To automate the process of creating native modules or views we need to implement a few things.
+First we need to introduce a new feature to the `create-react-native-library` package to create local modules. `create-react-native-library` will detect if the command has been run under React Native project if yes, it will create a local module under `local_modules` directory. CRNL will contain templates for creating native modules and views, also templates for turbo modules and fabric views.
+
+Second thing is to expand the abilities of autolinking to also link local modules. React Native CLI will link modules placed under `local_modules` directory and CLI will re-use logic that we have right now for adding native dependencies. See [here](https://github.com/react-native-community/cli/blob/main/docs/dependencies.md) for more references.
+
+Last thing is to adjust codegen script in React Native Core to also look under `local_modules` directory, because right now it looks under `node_modules` and dependencies specified in `react-native.config.js`.
+
+With these few things implemented - process of creating local modules will look:
+
+1. Run `npx create-react-native-library` under React Native app, to create module,
+2. Rebuild app. (install pods in case of iOS),
+3. Import module:
+
+    ```jsx
+    import React from 'react';
+    import { View } from 'react-native';
+    import helloWorld from './local_modules/hello_world'; // import module
+
+    helloWorld(); // call module
+
+    const App = () => {
+      return <View />
+    };
+
+    export default App;
+    ```
+
+Project structure will look:
+
+```
+- App.tsx
+- package.json
+- local_modules
+- hello_world
+  - hello-world.podspec
+  - index.ts (uses React Native APIs to exports module / view)
+  - ios
+    # native module code
+  - android
+    # native module code
+```
+
+Putting the files under a single directory under CRNL governance, makes it straight-forward to convert to a reusable library, and increases the chances for the module authors to open source their efforts. Separating all the local modules via different folders in one place will also make it easier to organize them and track the changes.
+
+## Alternatives
+
+- [react-native-coco-loco](https://github.com/jamonholmgren/react-native-colo-loco), it's a great solution - but it integrates with native code, proposed solution will autolink local modules and won't touch native code at all - which makes upgrades simpler and faster.
+- [Expo Local Modules](https://docs.expo.dev/modules/get-started/#creating-the-local-expo-module) Expo has a similar autolinking directive for modules that are using Expo Modules API.
```

Comment:
As you mentioned, this approach looks inspired from Local Expo Modules. 

Can we get someone from Expo to chime in here and help design this solution? (example: the fact that the folder on Expo is called `modules` and on your proposal is called `local_modules` is an unnecessary misalignment, to name one thing).

Reply:
Just wanted to clarify that I had no idea that about expo modules before we came up with the proposal. But yeah, we could align the naming if they behave similarly, e.g. expo modules would work with auto-linking in the CLI.